### ISSUE TITLE
是否可以提供baseline的实现代码呢？

### ISSUE THREAD

**Issue by youyuanyi** [yifan123/flow_grpo#37]
您好，您的Flow-GRPO工作对我们的研究非常有启发。我注意到您的论文中有和Flow-DPO作比较，我没有在Flow-DPO作者提供的code中找到相关实现。请问您能否提供Flow-DPO的代码呢。

**Comment by yifan123** [comment]
Thank you very much for your interest in our work. I will finish organizing the latest baseline code for SFT, RWR, and DPO, including both the online and offline versions, within the next two days.

**Comment by youyuanyi** [comment]
> Thank you very much for your interest in our work. I will finish organizing the latest baseline code for SFT, RWR, and DPO, including both the online and offline versions, within the next two days.

Thank you very much!

**Comment by yifan123** [comment]
Please use this entry point: [dpo.sh](https://github.com/yifan123/flow_grpo/blob/main/scripts/single_node/dpo.sh)

**Comment by youyuanyi** [comment]
> Please use this entry point

<img width="366" height="373" alt="Image" src="https://github.com/user-attachments/assets/98ee19a5-882e-431d-a0b6-2c4197b0608b" />

<img width="382" height="409" alt="Image" src="https://github.com/user-attachments/assets/580540be-ac74-4165-84d6-7fab78d697d6" />

<img width="382" height="381" alt="Image" src="https://github.com/user-attachments/assets/a9ba7412-cd78-4556-85d8-56839f505c49" />
Thank you very much for releasing the baseline code. I also have a quick question I'd like to ask.

I've reproduced FlowDPO and applied it to a portrait beauty enhancement task. The three images below are the results of fine-tuning LoRA on the FluxKontext model.

It seems that the model is able to distinguish between the chosen and rejected samples. However, in my experiments, the actual editing quality did not show noticeable improvement.

Could this be because the difference between my chosen and rejected samples is too small?

**Comment by youyuanyi** [comment]
> > Please use this entry point
> 
> <img alt="Image" width="366" height="373" src="https://private-user-images.githubusercontent.com/67009919/466848980-98ee19a5-882e-431d-a0b6-2c4197b0608b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI2NDg0ODAsIm5iZiI6MTc1MjY0ODE4MCwicGF0aCI6Ii82NzAwOTkxOS80NjY4NDg5ODAtOThlZTE5YTUtODgyZS00MzFkLWEwYjYtMmM0MTk3YjA2MDhiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE2VDA2NDMwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5NmNiY2QyZDk1NTlkMWE4ODViOGZiYWY5M2M1MjRiZjkxZjUyZDU2ODY4MzhmNjQ3YjZjZDVhMGJlNzk1YzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CMqRYT7S8ypTMEO9SMNQmsTM1uVbE6v3xJgBQGSruT8"> <img alt="Image" width="382" height="409" src="https://private-user-images.githubusercontent.com/67009919/466849177-580540be-ac74-4165-84d6-7fab78d697d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI2NDg0ODAsIm5iZiI6MTc1MjY0ODE4MCwicGF0aCI6Ii82NzAwOTkxOS80NjY4NDkxNzctNTgwNTQwYmUtYWM3NC00MTY1LTg0ZDYtN2ZhYjc4ZDY5N2Q2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE2VDA2NDMwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhNzVhYjZlYmI5MWMyZGFkYjY1NWI5ODI4MjhjNTU2NzlhN2RhYzRlNzVmNTljZDFmYzdiMmVmZWNlZjkzNTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.QH8ysvA9iOTNhp3OKstTSPxnJD96inGtU983NmjS9Jw"> <img alt="Image" width="382" height="381" src="https://private-user-images.githubusercontent.com/67009919/466849241-a9ba7412-cd78-4556-85d8-56839f505c49.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI2NDg0ODAsIm5iZiI6MTc1MjY0ODE4MCwicGF0aCI6Ii82NzAwOTkxOS80NjY4NDkyNDEtYTliYTc0MTItY2Q3OC00NTU2LTg1ZDgtNTY4MzlmNTA1YzQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE2VDA2NDMwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmNmE3NTliZTM4MmEyMTBkNzY5Yjk4MmE5MGM2YjBlMzg3NGNjNDEyOTQwOTg2MjVjODQxY2IwM2ZmNzYzNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JweYgoyRKTRHrdKc8H1Nvwlwf90VImBnkVCHkk04hn4">
> Thank you very much for releasing the baseline code. I also have a quick question I'd like to ask.
> 
> I've reproduced FlowDPO and applied it to a portrait beauty enhancement task. The three images below are the results of fine-tuning LoRA on the FluxKontext model.
> 
> It seems that the model is able to distinguish between the chosen and rejected samples. However, in my experiments, the actual editing quality did not show noticeable improvement.
> 
> Could this be because the difference between my chosen and rejected samples is too small?

Here is my code:

         `model_win_pred = transformer(
                    hidden_states=concat_win_packed,
                    # YiYi notes: divide it by 1000 for now because we scale it by 1000 in the transformer model (we should not keep it but I want to keep the inputs same for the model for testing)
                    timestep=timesteps / 1000,
                    guidance=guidance,
                    pooled_projections=pooled_prompt_embeds,
                    encoder_hidden_states=prompt_embeds,
                    txt_ids=text_ids,
                    img_ids=img_win_ids,
                    return_dict=False,
                )[0]

          model_lose_pred = transformer(
              hidden_states=concat_lose_packed,
              # YiYi notes: divide it by 1000 for now because we scale it by 1000 in the transformer model (we should not keep it but I want to keep the inputs same for the model for testing)
              timestep=timesteps / 1000,
              guidance=guidance,
              pooled_projections=pooled_prompt_embeds,
              encoder_hidden_states=prompt_embeds,
              txt_ids=text_ids,
              img_ids=img_lose_ids,
              return_dict=False,
          )[0]
          # ref model Predict the niose redisual
          with torch.no_grad():
              ref_win_pred = ref_transformer(
                  hidden_states=concat_win_packed,
                  # YiYi notes: divide it by 1000 for now because we scale it by 1000 in the transformer model (we should not keep it but I want to keep the inputs same for the model for testing)
                  timestep=timesteps / 1000,
                  guidance=guidance,
                  pooled_projections=pooled_prompt_embeds,
                  encoder_hidden_states=prompt_embeds,
                  txt_ids=text_ids,
                  img_ids=img_win_ids,
                  return_dict=False,
              )[0]

              ref_lose_pred = ref_transformer(
                  hidden_states=concat_lose_packed,
                  # YiYi notes: divide it by 1000 for now because we scale it by 1000 in the transformer model (we should not keep it but I want to keep the inputs same for the model for testing)
                  timestep=timesteps / 1000,
                  guidance=guidance,
                  pooled_projections=pooled_prompt_embeds,
                  encoder_hidden_states=prompt_embeds,
                  txt_ids=text_ids,
                  img_ids=img_lose_ids,
                  return_dict=False,
              )[0]
          
          model_win_pred = model_win_pred[:, :packed_noisy_win_input.size(1)] 
          model_win_pred = FluxKontextPipeline._unpack_latents(
              model_win_pred,
              height=latent_edited.shape[2] * vae_scale_factor,
              width=latent_edited.shape[3] * vae_scale_factor,
              vae_scale_factor=vae_scale_factor,
          ) # [bsz,16,128,128]

          model_lose_pred = model_lose_pred[:, :packed_noisy_lose_input.size(1)] 
          model_lose_pred = FluxKontextPipeline._unpack_latents(
              model_lose_pred,
              height=latent_reject.shape[2] * vae_scale_factor,
              width=latent_reject.shape[3] * vae_scale_factor,
              vae_scale_factor=vae_scale_factor,
          ) # [bsz,16,128,128]


          ref_win_pred = ref_win_pred[:, :packed_noisy_win_input.size(1)] 
          ref_win_pred = FluxKontextPipeline._unpack_latents(
              ref_win_pred,
              height=latent_edited.shape[2] * vae_scale_factor,
              width=latent_edited.shape[3] * vae_scale_factor,
              vae_scale_factor=vae_scale_factor,
          ) # [bsz,16,128,128]

          ref_lose_pred = ref_lose_pred[:, :packed_noisy_lose_input.size(1)] 
          ref_lose_pred = FluxKontextPipeline._unpack_latents(
              ref_lose_pred,
              height=latent_reject.shape[2] * vae_scale_factor,
              width=latent_reject.shape[3] * vae_scale_factor,
              vae_scale_factor=vae_scale_factor,
          ) # [bsz,16,128,128]
         target_win = noise - latent_edited
         target_lose = noise - latent_reject
         model_w_err = (model_win_pred  - target_win ).pow(2).mean(dim=[1,2,3])  # 形状 [B]
                model_l_err = (model_lose_pred - target_lose).pow(2).mean(dim=[1,2,3])
                ref_w_err   = (ref_win_pred  - target_win ).pow(2).mean(dim=[1,2,3])
                ref_l_err   = (ref_lose_pred - target_lose).pow(2).mean(dim=[1,2,3])
           
                model_diff = model_w_err - model_l_err
                ref_diff = ref_w_err - ref_l_err
                raw_model_loss = 0.5 * (model_w_err.mean() + model_l_err.mean())

                t = timesteps.float()/noise_scheduler_copy.config.num_train_timesteps
                scale_term = -0.5 * args.beta_dpo
                inside_term = scale_term * (model_diff-ref_diff)
                implicit_acc = (inside_term>0).sum().float() / inside_term.size(0)
                # ε̂_w − ε̂_w^ref
                win_pred_diff = model_win_pred - ref_win_pred
                
                model_w_pre_err = model_win_pred  - target_win
                loss = -1 * F.logsigmoid(inside_term).mean()
`

**Comment by yifan123** [comment]
It seems that the distinction between the chosen and rejected responses is very clear — the implicit accuracy reaches 100% after a short period of training. Could you please share how you constructed your chosen and rejected data?

**Comment by youyuanyi** [comment]
> It seems that the distinction between the chosen and rejected responses is very clear — the implicit accuracy reaches 100% after a short period of training. Could you please share how you constructed your chosen and rejected data?

We constructed a beauty enhancement dataset consisting of input images (before beautification) and manually enhanced label images (after beautification).

We fine-tuned the FluxKontext model using LoRA on this dataset. However, the fine-tuned model still fails to match the target label quality in terms of skin smoothing naturalness, makeup preservation, and other fine-grained details.

Therefore, we treat the model-generated images as reject samples, and their corresponding label images as chosen samples.

This allows us to construct a dataset in the form of:
<prompt, reject: images generated by LoRA, chosen: label images>,
which is used to supervise preference-based alignment.
